not-script/not-script.c

```diff
@@ -0,0 +1,529 @@
+#define _GNU_SOURCE 1
+#include <inttypes.h>
+#include <limits.h>
+#include <stddef.h>
+#include <assert.h>
+#include <stdbool.h>
+#include <errno.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <sys/ioctl.h>
+#include <sys/sysmacros.h>
+#include <sys/stat.h>
+#include <fcntl.h>
+#include <unistd.h>
+
+#include <err.h>
+
+#include <linux/major.h>
+#include <linux/loop.h>
+#include <linux/fs.h>
+
+#include <xen/xen.h>
+#include <xen/io/xenbus.h>
+#include <xenstore.h>
+
+static int open_file(const char *path) {
+    return open(path, O_RDONLY | O_NOCTTY | O_CLOEXEC | O_NOFOLLOW);
+}
+
+/* A simple library for loop device ioctls */
+struct loop_context {
+    uint32_t fd;
+};
+
+static int open_loop_dev(uint32_t devnum, bool writable)
+{
+    char buf[sizeof("/dev/loop") + 10];
+    if ((unsigned)snprintf(buf, sizeof buf, "/dev/loop%u", (unsigned)devnum) >= sizeof buf)
+        abort();
+    int res = open(buf, (writable ? O_RDWR : O_RDONLY) |
+                   O_EXCL | O_CLOEXEC | O_NOCTTY | O_NOFOLLOW);
+    return res == -1 ? -errno : res;
+}
+
+static int setup_loop(struct loop_context *ctx,
+                      uint32_t fd,
+                      uint64_t offset,
+                      uint64_t sizelimit,
+                      bool writable, bool autoclear) {
+    int dev_file = -1, status;
+
+    struct loop_config config = {
+        .fd = fd,
+        .block_size = 0, /* FIXME! */
+        .info = {
+            .lo_offset = offset,
+            .lo_sizelimit = sizelimit,
+            .lo_flags = LO_FLAGS_DIRECT_IO | (autoclear ? LO_FLAGS_AUTOCLEAR : 0) | (writable ? 0 : LO_FLAGS_READ_ONLY),
+        },
+    };
+
+    for (int retry_count = 0; retry_count < 5 /* arbitrary */; retry_count++) {
+        if ((status = ioctl(ctx->fd, LOOP_CTL_GET_FREE)) == -1)
+            return -errno;
+        dev_file = open_loop_dev(status, writable);
+        if (dev_file > 0) {
+            switch (ioctl(dev_file, LOOP_CONFIGURE, &config)) {
+            case 0:
+                return dev_file;
+            case -1:
+                if (close(dev_file))
+                    abort(); /* cannot happen on Linux */
+                return -1;
+            default:
+                abort();
+            }
+        }
+        if (errno != EBUSY && errno != ENXIO)
+            break;
+    }
+    return -1;
+}
+
+#define DEV_MAPPER "/dev/mapper/"
+#define DEV_MAPPER_SIZE (sizeof DEV_MAPPER - 1)
+
+static void
+process_blk_dev(int fd, const char *path, bool writable, dev_t *dev,
+                uint64_t *diskseq, bool permissive, bool autoclear)
+{
+    struct stat statbuf;
+    char buf[45];
+
+    if (fstat(fd, &statbuf))
+        err(1, "fstat");
+
+    if (S_ISBLK(statbuf.st_mode)) {
+        if (permissive)
+            goto skip;
+
+        /* block device */
+        if (strncmp(DEV_MAPPER, path, DEV_MAPPER_SIZE))
+            errx(1, "Only device-mapper block devices are supported (got %s)", path);
```

Comment:
Tests are still running, but I see already this breaking several use cases, including most of `qvm-block`. There are a few solutions to this problem, including using this not-script only for few known-supported cases. But I'd prefer you adding support for arbitrary block devices (without any fancy extra handling, just write its major:minor to xenstore). Including device-mapper devices given by a different path (like `/dev/dm-0`).